R/derive_vars_cat.R

```diff
@@ -0,0 +1,218 @@
+#' Derive Categorization Variables Like `AVALCATy` and `AVALCAyN`
+#' @param dataset
+#' `r roxygen_param_dataset(expected_vars = c("by_vars", "definition"))`
+#' @param definition List of expressions created by: `exprs()`.
+#' Must be in the format of a tribble.
+#' Must contain:
+#'  - the column `condition` which evaluates to a logic in `dataset`.
+#'  - at least one additional column with the new column name and the category value.
+#'  - the column specified in `by_vars` (if `by_vars` is specified)
+#'
+#' e.g. if `by_vars` is not specified:
+#' `exprs(~condition, ~AVALCAT1, ~AVALCA1N
+#'        AVAL >= 140, ">=140 cm", 1,
+#'        AVAL < 140, "<140 cm", 2)`
+#'
+#' e.g. if `by_vars` is specified as `exprs(VSTEST)`:
+#' `exprs(~VSTEST, ~condition, ~AVALCAT1, ~AVALCA1N
+#'        "Height", AVAL >= 140, ">=140 cm", 1,
+#'        "Height", AVAL < 140, "<140 cm", 2)`
+#'
+#' @param by_vars list of expressions with one element. `NULL` by default.
+#' Allows for specifying by groups, e.g. `exprs(PARAMCD)`.
+#' Variable must be present in both `dataset` and `definition`.
+#' The conditions in `definition` are applied only to those records that match `by_vars`.
+#' The categorization variables are set to NA for records
+#' not matching any of the by groups in `definition`.
+#'
+#'
+#' @details
+#' If conditions are overlapping, the row order of `definitions` must be carefully considered.
+#' The **first** match will determine the category.
+#' i.e. if
+#'
+#' `AVAL = 155`
+#'
+#' and the `definition` is:
+#'
+#' definition <- exprs(
+#'   ~VSTEST, ~condition, ~AVALCAT1, ~AVALCA1N,
+#'   "Height", AVAL > 170, ">170 cm", 1,
+#'   "Height", AVAL <= 170, "<=170 cm", 2,
+#'   "Height", AVAL <= 160, "<=160 cm", 3
+#' )
+#'
+#' then `AVALCAT1` will be `"<=170 cm"`, as this is the first match for `AVAL`.
+#' If you specify:
+#'
+#' definition <- exprs(
+#'   ~VSTEST, ~condition, ~AVALCAT1, ~AVALCA1N,
+#'   "Height", AVAL <= 160, "<=160 cm", 3,
+#'   "Height", AVAL <= 170, "<=170 cm", 2,
+#'   "Height", AVAL > 170, ">170 cm", 1
+#' )
```

Comment:
Possible to make as tables so renders a bit nicer? 

![image](https://github.com/user-attachments/assets/30dd73a3-c140-481a-8ae6-c371d4e3f206)

Reply:
Done